makefiles/tools/esptool.inc.mk

```
@@ -31,19 +32,38 @@ else
   FFLAGS += 0x10000 $(FLASHFILE).bin
 endif
 
-.PHONY: esp-image-convert esp-qemu
-# prepare image to flash: convert .elf to .bin
-esp-image-convert:
+# This target actually depends on $(FLASHFILE), but if we add $(FLASHFILE) here
+# it will cause it to be rebuilt even in "flash-only" mode because ELFFILE is
+# FORCEd. To avoid that we either add the dependency when calling make with
+# "all" or "flash", or we make it a PHONY target when only doing "flash-only".
+$(FLASHFILE).bin:
 	$(Q)$(ESPTOOL) --chip $(FLASH_CHIP) elf2image --flash_mode $(FLASH_MODE) \
 		--flash_size $(FLASH_SIZE)MB --flash_freq $(FLASH_FREQ) $(FLASH_OPTS) \
-		-o $(FLASHFILE).bin $(FLASHFILE)
+		-o $@ $(FLASHFILE)
+
+ifneq (,$(filter all flash, $(MAKECMDGOALS)))
+$(FLASHFILE).bin: $(FLASHFILE)
+else
+ifneq (,$(filter flash-only, $(MAKECMDGOALS)))
+.PHONY: $(FLASHFILE).bin
+endif
+endif
```

```suggestion
ifneq (,$(filter all flash, $(MAKECMDGOALS)))
$(FLASHFILE).bin: $(FLASHFILE)
else
ifneq (,$(filter flash-only, $(MAKECMDGOALS)))
.PHONY: $(FLASHFILE).bin
endif
endif
```

With all the above changes this would not be needed.

---

Done.